File: reference/src/reconciler/diff.ts
/**
 * Resource diff utilities for reconciler optimization.
 * Allows skipping reconciliation when only .status has changed (PRD Section 12).
 */

import { createHash } from 'node:crypto';
import type { AnyResource } from '../core/types.js';

/**
 * Compute a SHA-256 fingerprint of a resource's spec and metadata,
 * ignoring .status entirely. Used for O(1) change detection.
 */
export function resourceFingerprint(resource: AnyResource): string {
  const payload = JSON.stringify({
    spec: resource.spec,
    metadata: {
      name: resource.metadata.name,
      labels: resource.metadata.labels,
      annotations: resource.metadata.annotations,
    },
  });
  return createHash('sha256').update(payload).digest('hex');
}

/**
 * Check whether a resource's spec or metadata has changed between two snapshots.
 * Status-only changes return false (no spec change).
 */
export function hasSpecChanged(previous: AnyResource, current: AnyResource): boolean {
  return resourceFingerprint(previous) !== resourceFingerprint(current);
}

export interface ResourceCache {
  /** Returns true if the resource should be reconciled (spec/metadata changed or new). */
  shouldReconcile(resource: AnyResource): boolean;
  /** Clear all cached fingerprints. */
  clear(): void;
  /** Number of cached entries. */
  size(): number;
}

/**
 * Create a fingerprint cache for efficient change detection.
 * Resources are keyed by metadata.name. On first sight, always returns true.
 * On subsequent checks, only returns true if spec/metadata fingerprint changed.
 */
export function createResourceCache(): ResourceCache {
  const cache = new Map<string, string>();

  return {
    shouldReconcile(resource: AnyResource): boolean {
      const name = resource.metadata.name;
      const fingerprint = resourceFingerprint(resource);
      const cached = cache.get(name);

      if (cached === fingerprint) {
        return false;
      }

      cache.set(name, fingerprint);
      return true;
    },

    clear() {
      cache.clear();
    },

    size() {
      return cache.size;
    },
  };
}

File: sdk-go/reconciler/diff.go
package reconciler

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"sync"

	"github.com/ai-sdlc-framework/ai-sdlc/sdk-go/core"
)

// ResourceFingerprint computes a SHA-256 fingerprint of a resource's spec.
func ResourceFingerprint(resource core.AnyResource) (string, error) {
	data, err := json.Marshal(resource)
	if err != nil {
		return "", fmt.Errorf("failed to marshal resource: %w", err)
	}
	h := sha256.Sum256(data)
	return fmt.Sprintf("%x", h), nil
}

// ResourceCache tracks resource fingerprints for change detection.
type ResourceCache struct {
	mu           sync.RWMutex
	fingerprints map[string]string // key: "kind/name" -> fingerprint
}

// NewResourceCache creates a new resource cache.
func NewResourceCache() *ResourceCache {
	return &ResourceCache{fingerprints: make(map[string]string)}
}

func cacheKey(resource core.AnyResource) string {
	return string(resource.GetKind()) + "/" + resource.GetMetadata().Name
}

// HasSpecChanged returns true if the resource spec has changed since last check.
func (c *ResourceCache) HasSpecChanged(resource core.AnyResource) (bool, error) {
	fp, err := ResourceFingerprint(resource)
	if err != nil {
		return false, err
	}

	key := cacheKey(resource)

	c.mu.RLock()
	prev, exists := c.fingerprints[key]
	c.mu.RUnlock()

	if !exists || prev != fp {
		c.mu.Lock()
		c.fingerprints[key] = fp
		c.mu.Unlock()
		return true, nil
	}
	return false, nil
}
